### PR TITLE
TF: sample generation compatible with XLA and dynamic batch sizes

### LINKED ISSUE #19747: (TF) model.generate with beam search to an exportable tf.function
### Feature request

It would be good if we could export the `tf.function` `model.generate` with beam search, so it is servable with TF Serving. 

### Motivation

When we try to export a `model.generate` with `do_sample=True` to a `tf.function` and save it, we get an error due to Transformers and Tensorflow Internals. It would be good if, such as what is done in #18372, we could do for beam search:

```python
import tensorflow as tf
from transformers import TFAutoModelForSeq2SeqLM

class MyOwnModel(tf.Module):
    def __init__(self, model_path="t5-small"):
        super(MyOwnModel, self).__init__()
        self.model = TFAutoModelForSeq2SeqLM.from_pretrained(model_path)

    @tf.function(input_signature=(tf.TensorSpec((None, 32), tf.int32, name="input_ids"),
                                  tf.TensorSpec((None, 32), tf.int32, name="attention_mask")), jit_compile=True)
    def serving(self, input_ids, attention_mask):
        outputs = self.model.generate(input_ids=input_ids, attention_mask=attention_mask, max_new_tokens=32,
                                   return_dict_in_generate=True, do_sample=True)
        return {"sequences": outputs["sequences"]}

model = MyOwnModel()
export_dir = "./t5-example"

tf.saved_model.save(
    model,
    export_dir,
    signatures={
        "serving_default":
            model.serving
    })
```

That way, we would be able to serve TF text to text models with beam search in TF Serving, which would be a blast. Otherwise we would have to implement our own servers which is not ideal.

### Your contribution

I can try submitting a PR, but would be glad if I could get some guidance because this one seems harsh. It seems like we would have to remove lots of `shape` related operations.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes #19747 (sample generation failing with XLA and dynamic batch sizes)

There were two distinct problems:
- `_expand_inputs_for_generation` was crashing when the input was a `TensorSpec` with dynamic batch size (used e.g. with tf.Serving). Fix: `tensor.shape` -> `tf.shape`
- Obtaining a new seed with `tf.random.Generator` is broken with `tf.function` for two distinct reasons: 1) the protection we added [here](https://github.com/huggingface/transformers/pull/18044/files) makes unseeded `sample` crash unless we call `sample` with eager execution first (see [here](https://www.tensorflow.org/api_docs/python/tf/random/set_global_generator) why); 2) TF 2.10 + XLA crashes on the line that instantiates it. Fix: replace TF-suggested method to generate seed with `tf.random.Generator` with `tf.experimental.numpy.random.randint`, since we were creating a Generator from a non-deterministic state anyways 🤷 Notes: the new `tf.experimental.numpy.random.randint` strategy seems to be slightly faster (23.5 ms -> 21.6 ms per generate call on a benchmark with XLA, nvidia 3090,  and `t5-small`) and is retrocompatible up to TF 2.4.0.

### CODE REVIEW

File: src/transformers/generation_tf_utils.py [hunk 1]
```diff
@@ -2624,7 +2625,7 @@ def sample_body_fn(generated, finished_sequences, cur_len, model_kwargs):
             if seed is not None:
                 sample_seed = seed
             else:
-                sample_seed = tf.cast(self.seed_generator.make_seeds(count=1)[:, 0], dtype=tf.int32)
+                sample_seed = tf.experimental.numpy.random.randint(tf.int32.min, tf.int32.max, (2,), dtype=tf.int32)
```
Comment (sgugger) [inline src/transformers/generation_tf_utils.py:2628]: Do we really have to rely on tf.experimental here? (e.g. it's not possible to convert this from a numpy array?)

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

File: src/transformers/generation_tf_utils.py [hunk 2]
```diff
@@ -360,6 +360,7 @@ class TFGenerationMixin:
 
     @property
     def seed_generator(self):
+        warnings.warn("`seed_generator` is deprecated and will be removed in a future version.", UserWarning)
```
Comment (gante) [inline src/transformers/generation_tf_utils.py:363]: @sgugger `seed_generator` is public and therefore can't be removed, correct? (despite being recent and exclusively used in `sample`)

Comment (sgugger) [reply src/transformers/generation_tf_utils.py:363]: A deprecation cycle is nicer yes. We can consider removing it in a couple of minors if it was super recent

File: src/transformers/generation_tf_utils.py (hunk 1 above)
Comment (gante) [reply src/transformers/generation_tf_utils.py:2628]: We can convert from a numpy array, but the execution speed penalty with XLA is huge (`t5-small` went from ~23ms/generation to ~56ms/generation on my machine). Sadly this is the only TF-based alternative to `tf.random.Generator` strategies

`tf.experimental.numpy` has been pretty stable since TF 2.4 🤔 

@Rocketknight1 what are your thoughts here?

Comment (Rocketknight1) [reply src/transformers/generation_tf_utils.py:2628]: `tf.experimental.numpy` is extremely non-experimental by now, it's been around since TF 2.4 and is very unlikely to change significantly! I'm totally fine with using it across the codebase.

**Review by Rocketknight1 — APPROVED** [review]
LGTM! As I mentioned, I don't see any issue with using `tf.experimental.numpy` given that it's been stable for almost 2 years now.
